### PR TITLE
[SPARK-48784][SQL] Add ::: syntax as a shorthand for try_cast

### PR DESCRIPTION
**Pull request by gene-db**
### What changes were proposed in this pull request?

Add the `:::` (triple colon) as syntactic sugar for `try_cast`.

### Why are the changes needed?

This syntactic sugar makes it easier/simpler to call `try_cast`. Now, something like this is possible:

```
> select '123.456':::int;
  NULL
```

### Does this PR introduce _any_ user-facing change?

Yes, this is new syntax. It enables using `:::` as shorthand for `try_cast`.

### How was this patch tested?

New unit tests and sql tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by HyukjinKwon — DISMISSED** [review]

**Comment by yaooqinn** [comment]
It looks very uncommon and incurs a cognitive cost for users. Also cc @dongjoon-hyun

**Comment by srielau** [comment]
> It looks very uncommon and incurs a cognitive cost for users. Also cc @dongjoon-hyun

It is uncommon because it is new.
Note that other DBMS add short hands as well.
The first question is whether SparkSQL can innovate on language, or it must always follow.
If we agree that SparkSQL can innovate, then we can discuss the way to achieve the goal.
Given that cast <==> :: (originally introduced in the 1990's (??) by Informix - I think.
It seems logical to play on `::`.
A single `:` is already used for JSON path.

**Comment by dongjoon-hyun** [comment]
I'm also reluctant like @yaooqinn 's comment.

Could you summarize the status of other DBMS in the PR description first for further discussion?

**Comment by srielau** [comment]
> I'm also reluctant like @yaooqinn 's comment.
> 
> Could you summarize the status of other DBMS in the PR description first for further discussion?

As noted above SparkSQL would the first (that I know of) to introduce a shorthand for try_cast.
try_cast() is supported at least by MS SQL Server, IBM Db2 , Snowflake, duckDB, NOT PostgreSQL, NOT mySQL, it is NOT ANSI SQL.
SparkSQL introduced try_cast to compensate for non-ANSI CASTing.

`::` is supported at least by PostgreSQL, Snowflake, Informix, Db2, duckDB (It's hard to search the docs for "::"). It is also NOT ANSI

Personally I prefer `::` over CAST since it it so more dense to type and casts are frequently needed.

For that reason I think try_cast deserves a shorthand (which may or may not be `:::`) as well, I would not be surprised if duckDB copied Spark SQL if we lead the way in a sensible and obvious fashion.

**Comment by cloud-fan** [comment]
We need to be careful when adding new Spark-only SQL syntaxes, but it doesn't mean Spark can't be the first one to do innovation.

It seems to be a natural extension of the existing `::` syntax. Does it introduce any ambiguity to our parser rules? Where do we document it?

**Comment by mkaravel** [comment]
When Redshift introduced SUPER they faced the same issue, which is basically the clunkiness of writing an effective query in the presence of the SUPER data type, which (as the VARIANT type in OSS) implements dynamic typing. See the [corresponding documentation](https://docs.aws.amazon.com/redshift/latest/dg/query-super.html#dynamic-typing-lax-processing) for a relevant example.

Redshift’s solution was to embed try_cast semantics in all operators to avoid making users to write very/extremely complicated queries. The choice was obviously made to made to make users’ life easier (ease of use) and of course to make the resulting SQL code much more readable. One has to note here though that Redshift does not have try_cast functionality.

My (personal) take: the idea of having LAX (this is what Redshift calls it) semantics is actually paramount when we are talking about dynamic typing (which is what Redshift's SUPER and OSS VARIANT are trying to do), and makes usability of the new types orders of magnitude simpler.

The proposed syntax in OSS (again in my opinion) goes one step further than what Redshift does in the sense that we not only offer LAX semantics (which are essential and necessary), but also does it in a user-controlled way. In our case (Spark) the alternative would have been to populate try_cast calls to the entire query. I believe the proposed syntax makes it much more simple, and in IMHO much more intuitive (use of a cast operator when a cast is needed as opposed to a function call).

So bottom line: Spark is not the first one to associate LAX semantics when it comes to dynamic/polymorphic data types, but with this PR OSS will the first one to make it part of the syntax in a nice manner that extends existing and established casting syntax (I am referring to `::`).

**Comment by srielau** [comment]
> We need to be careful when adding new Spark-only SQL syntaxes, but it doesn't mean Spark can't be the first one to do innovation.
> 
> It seems to be a natural extension of the existing `::` syntax. Does it introduce any ambiguity to our parser rules? Where do we document it?

I don't think there can be an ambiguity.
It would require`:: :param` which is not legal.

As for docs: You are making a good point. I cannot find existing docs for `try_cast` or `::`. Should fix all that in one go.

**Comment by mkaravel** [comment]
> It looks very uncommon and incurs a cognitive cost for users. Also cc @dongjoon-hyun

When something new is added in a language there is naturally a cognitive overhead to adapt to the newly added feature. That alone is just one side of the coin. The other side is the added value of the new feature for the users of that language. 

The operator proposed in this PR allows to write queries in a most succinct and readable manner (there does not seem to be any ambiguity in the proposed syntax).

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I read all above comments. Although I understand and appreciate all of your passion, it seems that I need to cast my -1 explicitly at least for Apache Spark 4.0.0. For Apache Spark 4.0.0, we hope to be more closer to ANSI SQL with less esoteric syntaxes (like this PR). Sorry but do you think we can simply postpone this PR a few months and start to discuss after releasing 4.0.0 release successfully?

**Comment by cloud-fan** [comment]
@dongjoon-hyun Many databases have extensions over the SQL standard and so does Spark. I don't think delaying the discussion helps (BTW there is another one ongoing: https://github.com/apache/spark/pull/46707). We should discuss the value of the new syntax and accept it if the value is well justified.

**Comment by dongjoon-hyun** [comment]
> @dongjoon-hyun Many databases have extensions over the SQL standard and so does Spark. I don't think delaying the discussion helps (BTW there is another one ongoing: #46707). We should discuss the value of the new syntax and accept it if the value is well justified.

To @cloud-fan , like you say, this can be added as `SparkSessionExtensions`, not in Spark SQL built-in syntax, can't this?

**Comment by cloud-fan** [comment]
@dongjoon-hyun It's like why we add stored procedure into Spark. Iceberg can use extension APIs to implement the CALL command, but it's better to have it in Spark out of the box for ease of use. The same thing applies to many SQL features that are not SQL standard. We should have them in Spark out of the box if the value is well justified.

**Comment by srielau** [comment]
We are proposing a non-ANSI (`:::`) extension to a non-ANSI expression (`try_cast`) mimicking a non-ANSI extension (`::`) to an ANSI expression (`cast`). All motivated by ease of use for a non-ANSI type `VARIANT`.
Seems ANSI isn't really in play here. That ship sailed when we did `try_cast`, adopted `::` and added `VARIANT`.

**Comment by yaooqinn** [comment]
I also cast my vote against this because Spark is a platform built upon Scala, in which `:::` means the concatenation of two lists. So I don't see the value of adding this, but it would only incur cognitive costs for users.

**Comment by dongjoon-hyun** [comment]
To @cloud-fan , this PR is only adding a syntactic sugar, isn't it?
>  We should have them in Spark out of the box if the value is well justified.

For Iceberg `CALL` command, it was popular with Spark + Iceberg, but I'm not sure if this new syntax is already popular in Spark community. Maybe, do you mean Databricks has been providing this syntax already, @cloud-fan ?
- https://iceberg.apache.org/docs/1.5.1/spark-procedures/

**Comment by cloud-fan** [comment]
> ... because Spark is a platform built upon Scala

Spark has a lot more users on Python and SQL, I don't really buy it. And we already have `::`. The value of `:::` is the same as `::`: it's just a syntax surgery and can make users' SQL queries neater and more readable.

We will add this feature to Databricks SQL and we just want to share this nice feature with the community. If you don't agree, we will keep it in Databricks only.

**Comment by dongjoon-hyun** [comment]
It's good to know. If we need this in the Spark codebase, please provide a way for the community to free from this syntax in production environments. I believe that can be achievable as a built-in extension class instead of this kind of enforcement.
> We will add this feature to Databricks SQL and we just want to share this nice feature with the community. If you don't agree, we will keep it in Databricks only.

**Comment by cloud-fan** [comment]
Do you mean adding a config to fail this syntax by default? BTW if people write `:::` today they will get ANTLR parser errors.

**Comment by dongjoon-hyun** [comment]
> Do you mean adding a config to fail this syntax by default?

No~ I meant to provide this as a built-in extension to enable this syntax. And, by default, the extension and this new syntax is not supported.
https://github.com/apache/spark/blob/8ca1822b4f71145ebf7f61e4235a6fb9bb474805/sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensionsProvider.scala#L65-L70

**Comment by srielau** [comment]
To bypass the worry about scala: How about `?::` It is arguably more descriptive. And a leading `?` could be used for other operations like ` a ?/ b` (instead of try_divide(a, b))

**Comment by cloud-fan** [comment]
@dongjoon-hyun The parser extension is known to be cumbersome. We need to provide a fully functional ANTLR parser which will be run before Spark's own parser and fallback to Spark's parser if the extension parser can't parse. It usually means copy-paste many lexer/parser rules from the Spark codebase. It will be even harder to use once we have SQL scripting (which needs to parse things recursively). The extension parser basically needs to fork Spark's entire ANTLR parser and add its own syntax. If we agree to have this syntax in Spark, we should add it directly.

@srielau oh `?::` is a good idea! It can be applied to all operators that need a try version.
